Real Randomness

BroderPeters opened the pull request:
Löst das Random-Problem in der RandomAI.

Getestet.

Commit:
[#134] add Guid.NewGuid().GetHashCode() in Random(int Seed) as Seed

nicories commented:
Kann rein von mir aus. Das ist zufällig genug.

jmattheis commented:
Jain, es soll ja eine lösung sein, das es für alle ai zählt, das man zb schon ein random ding in der ai bereitstellt zb über die aibase

BroderPeters commented:
@jmattheis Einerseits ist das ja letzendlich dem Spieler überlassen, vielleicht SOLLEN die Ameisen ja parallel laufen.^^ Andererseits, gehört das dann nicht eher in den Helper-Ordner als eigene Random-Klasse? Ich meine wo nutzen wir sonst noch Random?

jmattheis commented:
Wenn die Ameisen Parallel laufen sollen kann er das selbst schreiben und nicht einen Bug benutzten das das Random object zur selben zeit erstellt wird und gleiche zahlen zurück gibt, was soll eine klasse im helper ding helfen?

nicories commented:
Eine separate Klasse fine ich nicht schlecht. Da kann dann das hier als Methode rein zum zufällig hin und her laufen und noch ein paar andere Hilfsmethoden.
`PathfindingUtils`

jmattheis commented:
`PathfindingUtils` das soll ja selbst manuell in der KI gemacht werden. Beispiele für methoden daraus?

BroderPeters commented:
Ich stimme Jannis zu. Pathfinding hat der Spieler selber zu schreiben.
Wie soll nun das Random-Problem gelöst werden?
Ich finde den Ansatz nicht falsch es dem Spieler bereitzustellen, da so wie ich es gemacht habe, es später keinerlei Auswirkung haben wird, da diese KI ja nur zu Testzwecken ist.

BroderPeters commented:
@jmattheis Laut Dokumentation finde ich aber die AIBase als eher unpassend:
`Die Basis der AI welcher Ameisen kaufen kann.`
`public abstract class AIBase : IAI {`

Dann eher:

`Die Basis für die KI der Ameise.`
`public abstract class AIAntBase : IAIAnt {`

jmattheis commented:
> Wie soll nun das Random-Problem gelöst werden?

Wie ich oben schon geschrieben habe:

> das man zb schon ein random ding in der ai bereitstellt zb über die aibase

das man dann eine `public Random getRandom();` Methode hat, oder es auch schon als variable in der AIBase deklariert wird und man die dann einfach nutzen kann.

jmattheis commented:
Edit meine natürlich die `AIAntBase` ^^

Commit:
[#134] add method getRandom() to get a instance of a random-class with a seed in AIAntBase, add method call getRandom() in Player1AI/AI.cs, updated RandomAI.dll

jmattheis commented on Player1AI/AI.cs:
```diff
@@ -33,6 +32,7 @@ public class AIAnt : AIAntBase {
             // Zucker aufheben Test
             Ant.pickUpSugar();
             // RANDOM FTW
+            Random rand = getRandom();
```
Bei jedem AntTick ein neues Random object zu erstellen ist unperformant.

BroderPeters replied:
So gebe er mir einen besseren Vorschlag.

jmattheis commented on AntWars/AI/AIAntBase.cs:
```diff
@@ -33,5 +33,13 @@ public abstract class AIAntBase : IAIAnt {
         public abstract void notify(HashSet<Coordinates> coords);
+
+        /// <summary>
+        /// Initialisiert eine neue Instanz der Random-Klasse unter Verwendung eines Startwerts.
```
`unter Verwendung eines Startwerts.` ? du setzt einen Seed aber keinen Startwert.

BroderPeters replied:
Habe es von der [maschinell übersetzten offiziellen MSDN-Seite](https://msdn.microsoft.com/de-de/library/ctssatww%28v=vs.110%29.aspx)

jmattheis replied:
Seed - Eine Zahl, **mit** der ein Startwert für Folgen von Pseudozufallszahlen berechnet wird.

Der Seed ist nicht der Startwert.

Commit:
change Startwert in summary of getRandom() to Seed

jmattheis commented on AntWars/AI/AIAntBase.cs:
```diff
@@ -33,5 +33,13 @@ public abstract class AIAntBase : IAIAnt {
         public abstract void notify(HashSet<Coordinates> coords);
+
+        /// <summary>
+        /// Initialisiert eine neue Instanz der Random-Klasse unter Verwendung eines Seeds.
```
Will jetzt ja nicht pingelig sein, aber wenn man kein Seed angibt wird trotzdem einer benutzt. 

Hab noch eine bessere Idee für die Implementation bekommen. Wir brauchen eigentlich nicht für jede AI ein eigenes Random Object, das ist ja theoretisch nur Speicher verbrauch deswegen sollte es innerhalb der AI nur 1ne Random instance geben, Dadurch brauchen wir garkein eigenen Seed berechnen.

```
private static readonly Random randomInstance = new Random();
public Random getRandom() {
    return randomInstance;
}
```

BroderPeters replied:
Also füge ich der Klasse `AI.cs` das Attribute `randomInstance` hinzu oder habe ich da wieder etwas falsch verstanden?

jmattheis replied:
`AIAntBase`, eigentlich nur das oben kopieren (:

Commit:
[#134] changes AIAntBase#getRandom()

AIAntBase#getRandom() returns now an allready existing instance of the class random

jmattheis commented on AntWars/AI/AIAntBase.cs:
```diff
@@ -35,11 +37,12 @@ public abstract class AIAntBase : IAIAnt {
-        public Random getRandom() {
-            return new Random(Guid.NewGuid().GetHashCode());
+        public Random getRandom()
+        {
```
formatting?, du kannst zwar selbst mit vs2013 die datei nicht öffnen aber du kannst da auch einstellungen drin machen das die klammern mit in der methoden deklarierung sind.

BroderPeters replied:
Sorry, habe ich übersehen.

Hatt es auch eigentlich 1:1 kopiert.

jmattheis commented on AntWars/AI/AIAntBase.cs:
```diff
@@ -20,6 +20,8 @@ public abstract class AIAntBase : IAIAnt {
 
         internal Config Conf { get; set; }
 
+        private static readonly Random randomInstance = new Random();
```
static /constante attribute über den "normalen" variablen (nice to have)

BroderPeters replied:
Hatte es wegen `private` unter internal eingeordnet.

Hatten wir nicht in irgend einem Issue ne Konvention zu Anordnung?

jmattheis replied:
Ich glaube das hatte ich bisjetzt immer nur in reviews geschrieben

Commit:
[#134] formatting

BroderPeters commented:
Kann der gemergt werden?

jmattheis commented:
Kann gemerged werden: Benutzt dann mal das GitHub [Squash Commits Feature](https://github.com/blog/2141-squash-your-commits) damit wir auf master nicht soviele unnötige commits haben

BroderPeters commented:
> Repository administrators now have a few options to choose from when deciding how to handle history.

EDIT:

> Users with admin permissions can configure pull request merge options on branches in several ways:
>     Allow both merge commits and commit squashing
>     Allow only merge commits
>     Allow only squash commits

jmattheis commented:
Jep, da kann man festlegen ob man blosnoch das eine oder das andere machen kann, du kannst momentan selbst auswählen was du machen willst in dem auf merge drückst und dann bei dem dropdown das auswählst
